examples/step-37/doc/results.dox

```diff
@@ -409,3 +412,232 @@ number of threads in the MPI_InitFinalize data structure in the main function,
+left hand side of the equation to finalize the right hand side
+contribution. Similar mathematics are also involved when first writing all
+entries into a left hand side matrix and then eliminating matrix rows and
+columns by MatrixTools::apply_boundary_values().
+
+In principle, the components that belong to the constrained degrees of freedom
+could be eliminated from the linear system because they do not carry any
+information. In practice, in deal.II we always keep the size of the linear
+system the same to avoid handling two different numbering systems and avoid
+confusion about the two different index sets. In order to ensure that the
+linear system does not get singular when not adding anything to constrained
+rows, we then add dummy entries to the matrix diagonal that are otherwise
+unrelated to the real entries.
+
+In a matrix-free method, we need to take a different approach, since the @p
+LaplaceOperator class represents the matrix-vector product of a
+<b>homogeneous</b> operator (the left-hand side of the last formula).  It does
+not matter whether the ConstraintMatrix passed to the MatrixFree::reinit()
+contains inhomogeneous constraints or not, the MatrixFree::cell_loop() call
+will only resolve the homogeneous part of the constraints as long as it
+represents a <b>linear</b> operator.
+
+In our matrix-free code, the right hand side computation where the
+contribution of inhomogeneous conditions ends up is completely decoupled from
+the matrix operator and handled by a different function above. Thus, we need
+to explicity generate the data that enters the right hand side rather than
+using a byproduct of the matrix assembly. Since we already know how to apply
+the operator on a vector, we could try to use those facilities for a vector
+where we only set the Dirichlet values:
+@code
+  // interpolate boundary values on vector solution
+  std::map<types::global_dof_index, double> boundary_values;
+  VectorTools::interpolate_boundary_values(mapping, dof_handler, 0,
+                                           BoundaryValueFunction<dim>(), boundary_values);
+  for (typename std::map<types::global_dof_index, double>::iterator it = boundary_values.begin();
+       it != boundary_values.end(); ++it)
+    if (solution.locally_owned_elements().is_element(it->first))
+      solution(it->first) = it->second;
+@endcode
+or, equivalently, if we already had filled the inhomogeneous constraints into
+a constraint matrix object,
+@code
+  solution = 0;
+  constraints.distribute(solution);
+@endcode
+
+We could then pass the vector @p solution to the @p LaplaceOperator::vmult_add()
+function and add the new contribution to the @p system_rhs vector that gets
+filled in the @p LaplaceProblem::assemble_rhs() function. However, this idea does
+not work because the vmult() functions eliminates the entries filled into
+Dirichlet-constrained degrees of freedom on the fly by construction. Instead,
+we could select one of two following strategies.
```

Comment:
i think the already apply internally in `project()` a third way: get RHS contributions from constraints by doing vmult, add to RHS and then through that into solver:
https://github.com/dealii/dealii/blob/master/include/deal.II/numerics/vector_tools.templates.h#L1304-L1326
This does not appear to be any of the two options below.

Reply:
i went through the two options now and I think what we do in `project_matrix_free` (originally written by @masterleinad ) is a more concise correct third option.

Reply:
I also checked this (and I remember having reviewed it), but it does not work. I just tried to change `ZeroFunction` into `ConstantFunction<dim>(1.)` in step-37 and it does ignore that. The reason is explained right in this block: The standard implementation of the `vmult()` method does set the constrained degrees of freedom to zero while reading them out of the vector. Thus, in absence of hanging node constraints it simply ignores those. I'm not sure how well we tested the `project()` function but I don't understand how it can work. The "correct" way of doing the `project` is to use the second option I describe.

Reply:
ok, but I still have a third options as a suggestion: use two different instances of `LaplaceOperator` and `MatrixFree` objects, one with constraints and one without. To me the advantage is that you can use exactly the same code in `vmult` etc, so no need to get special implementation only to make BC working.

Reply:
But isn't that just a variant of option two? (Except that I suggest to add a second `DoFHandler` object and create a second `LaplaceOperator` instance - it could of course also be an independent `MatrixFree` object. I can add it if you want there, or even show that as the "option 2" instead if you think it is easier to understand than adding a second DoFHandler/ConstraintMatrix pair.

Reply:
I though it could be more general if the system to be solved is block system. Then presumably you already reserve blocks for different blocks in the operator and it would be more difficult to introduce extra blocks only for BC.

I suggest to move the third option to `MatrixFreeOperators::Base` with a super short discussion and link to step-37.

Reply:
So how about I add a sentence in the second option that this could also be done via a separate `MatrixFree`/`LaplaceOperator` pair and then extend `MatrixFreeOperators::Base`?

Reply:
that sounds good to me.

Reply:
OK, I've added the discussion to `MatrixFreeOperators::Base`. Note that I changed the order of the two existing paragraphs to have the third option extend the second one (which is the selection of another block).